server/src/main/java/org/opensearch/search/approximate/ApproximateScoreQuery.java

```diff
@@ -0,0 +1,151 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.search.approximate;
+
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.search.BooleanClause;
+import org.apache.lucene.search.Explanation;
+import org.apache.lucene.search.IndexSearcher;
+import org.apache.lucene.search.Matches;
+import org.apache.lucene.search.Query;
+import org.apache.lucene.search.QueryVisitor;
+import org.apache.lucene.search.ScoreMode;
+import org.apache.lucene.search.Scorer;
+import org.apache.lucene.search.ScorerSupplier;
+import org.apache.lucene.search.Weight;
+
+import java.io.IOException;
+
+/**
+ * Base class for a query that can be approximated.
+ *
+ * This class is heavily inspired by {@link org.apache.lucene.search.IndexOrDocValuesQuery}. It acts as a wrapper that consumer two queries, a regular query and an approximate version of the same. By default, it executes the regular query and returns {@link Weight#scorer} for the original query. At run-time, depending on certain constraints, we can re-write the {@code Weight} to use the approximate weight instead.
+ */
+public final class ApproximateScoreQuery extends Query {
+
+    private final Query originalQuery, approximationQuery;
+
+    private Weight originalQueryWeight, approximationQueryWeight;
+
+    public ApproximateScoreQuery(Query originalQuery, Query approximationQuery) {
+        this(originalQuery, approximationQuery, null, null);
+    }
+
+    public ApproximateScoreQuery(
+        Query originalQuery,
+        Query approximationQuery,
+        Weight originalQueryWeight,
+        Weight approximationQueryWeight
+    ) {
+        this.originalQuery = originalQuery;
+        this.approximationQuery = approximationQuery;
+        this.originalQueryWeight = originalQueryWeight;
+        this.approximationQueryWeight = approximationQueryWeight;
+    }
+
+    public Weight getApproximationQueryWeight() {
+        return approximationQueryWeight;
+    }
+
+    public Query getOriginalQuery() {
+        return originalQuery;
+    }
+
+    public Query getApproximationQuery() {
+        return approximationQuery;
+    }
+
+    @Override
+    public Weight createWeight(IndexSearcher searcher, ScoreMode scoreMode, float boost) throws IOException {
+        originalQueryWeight = originalQuery.createWeight(searcher, scoreMode, boost);
+        approximationQueryWeight = approximationQuery.createWeight(searcher, scoreMode, boost);
```

Comment:
why are we creating the instances of `approximationQueryWeight ` and `approximationQueryScoreSupplier` when they are not getting used anywhere? 
It seems like implementations of `ApproximateScoreQuery` will have to override `createWeight()` anyways.

Reply:
So when a new `ApproximateScoreQuery` is initialized with both the `originalQuery` and the `approximationQuery`, we create and store their weights. When it is time for us to override the weight, we can simply set the query's weight to use this approximation query weight instead -- https://github.com/harshavamsi/OpenSearch/blob/612998e4b33c37ead73b7025a2b2d962f2538fa3/server/src/main/java/org/opensearch/search/internal/ContextIndexSearcher.java#L340

Reply:
I think we should avoid the need to create both weights by moving the call to `setContext` before the weight gets created. 

Maybe you could even move it to the start of `ContextIndexSearcher`'s `createWeight` implementation.

Then this method would just delegate to the correct subquery's `createWeight` method.